[quant] Enable 'same' and 'valid' padding in static qconv

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #78543
* #78386
* #77929
* __->__ #77928
* #77927

Resolves #76304

The string inputs in the quantized convolutions are not implemented, because:
1. Conv packing expects numerical padding as an input
2. The same padding computation is not implemented internally in the quantized conv

To resolve we use a commbination of padding and conv layers:

1. Precompute the padding for the `F.pad` (`_reverse_padding_repeat_twice`)
    1. If `padding == 'same'`: padding[i] = dilation[i] * (kernel[i] - 1)
    2. If `padding == 'valid'`: padding[i] = 0
1. If `padding_mode == 'zeros'` and `padding` is numeric
    1. Prepack the conv with self.padding
    1. Run the qconv in the forward as is
1. If `padding_mode != 'zeros'` or `padding` is a string
    1. In the constructor: Prepack the conv with 0
    1. In the forward:
        1. Run the `F.pad` with `_reversed_padding_repeated_twice`
        1. Run the qconv

The above steps *should not* affect the performance of the current implementaion,
but just add functionality for `'same'` and `'valid'` padding.

Test Plan:

```python
python test/test_quantzation.py TestInputOutputShapes
python test/test_quantzation.py TestStaticQuantizedModule
```

Differential Revision: [D36536582](https://our.internmc.facebook.com/intern/diff/D36536582)

Commit:
[quant] Enable 'same' and 'valid' padding in static qconv

Resolves #76304

The string inputs in the quantized convolutions are not implemented, because:
1. Conv packing expects numerical padding as an input
2. The same padding computation is not implemented internally in the quantized conv

To resolve we use a commbination of padding and conv layers:

1. Precompute the padding for the `F.pad` (`_reverse_padding_repeat_twice`)
    1. If `padding == 'same'`: padding[i] = dilation[i] * (kernel[i] - 1)
    2. If `padding == 'valid'`: padding[i] = 0
1. If `padding_mode == 'zeros'` and `padding` is numeric
    1. Prepack the conv with self.padding
    1. Run the qconv in the forward as is
1. If `padding_mode != 'zeros'` or `padding` is a string
    1. In the constructor: Prepack the conv with 0
    1. In the forward:
        1. Run the `F.pad` with `_reversed_padding_repeated_twice`
        1. Run the qconv

The above steps *should not* affect the performance of the current implementaion,
but just add functionality for `'same'` and `'valid'` padding.

[ghstack-poisoned]

z-a-f commented:
@z-a-f has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36536582).

Commit:
Update on "[quant] Enable 'same' and 'valid' padding in static qconv"


Resolves #76304

The string inputs in the quantized convolutions are not implemented, because:
1. Conv packing expects numerical padding as an input
2. The same padding computation is not implemented internally in the quantized conv

To resolve we use a commbination of padding and conv layers:

1. Precompute the padding for the `F.pad` (`_reverse_padding_repeat_twice`)
    1. If `padding == 'same'`: padding[i] = dilation[i] * (kernel[i] - 1)
    2. If `padding == 'valid'`: padding[i] = 0
1. If `padding_mode == 'zeros'` and `padding` is numeric
    1. Prepack the conv with self.padding
    1. Run the qconv in the forward as is
1. If `padding_mode != 'zeros'` or `padding` is a string
    1. In the constructor: Prepack the conv with 0
    1. In the forward:
        1. Run the `F.pad` with `_reversed_padding_repeated_twice`
        1. Run the qconv

The above steps *should not* affect the performance of the current implementaion,
but just add functionality for `'same'` and `'valid'` padding.

Test Plan:

```python
python test/test_quantzation.py TestInputOutputShapes
python test/test_quantzation.py TestStaticQuantizedModule
```

Differential Revision: [D36536582](https://our.internmc.facebook.com/intern/diff/D36536582)

[ghstack-poisoned]

z-a-f commented:
@z-a-f has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36536582).

Commit:
Update on "[quant] Enable 'same' and 'valid' padding in static qconv"


Resolves #76304

The string inputs in the quantized convolutions are not implemented, because:
1. Conv packing expects numerical padding as an input
2. The same padding computation is not implemented internally in the quantized conv

To resolve we use a commbination of padding and conv layers:

1. Precompute the padding for the `F.pad` (`_reverse_padding_repeat_twice`)
    1. If `padding == 'same'`: padding[i] = dilation[i] * (kernel[i] - 1)
    2. If `padding == 'valid'`: padding[i] = 0
1. If `padding_mode == 'zeros'` and `padding` is numeric
    1. Prepack the conv with self.padding
    1. Run the qconv in the forward as is
1. If `padding_mode != 'zeros'` or `padding` is a string
    1. In the constructor: Prepack the conv with 0
    1. In the forward:
        1. Run the `F.pad` with `_reversed_padding_repeated_twice`
        1. Run the qconv

The above steps *should not* affect the performance of the current implementaion,
but just add functionality for `'same'` and `'valid'` padding.

Test Plan:

```python
python test/test_quantzation.py TestInputOutputShapes
python test/test_quantzation.py TestStaticQuantizedModule
```

Differential Revision: [D36536582](https://our.internmc.facebook.com/intern/diff/D36536582)

[ghstack-poisoned]

z-a-f commented:
@z-a-f has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36536582).

vkuzo commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -17,7 +17,9 @@
     'zeros',
-    'reflect'
+    'reflect',
+    # 'circular',
```
nit: is this needed?

z-a-f replied:
Nope -- was testing  the support. Will remove it

vkuzo commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -79,6 +86,23 @@ def _init(self, in_channels, out_channels, kernel_size, stride,
+        # See `nn._ConvNd` for more details for the implementation.
+        # We have to reimplement it here, to support `torch.jit`
+        # TODO(z-a-f): Factor out common code
+        self._reversed_padding_repeated_twice = [0, 0] * len(self.kernel_size)  # 'valid' padding
```
this code seems very similar to the code in `torch.nn.modules.conv._ConvNd.__init__`, can we consider reusing instead of copying?

z-a-f replied:
I agree -- I tried doing that, but got `jit` issues that it's not serializable. I added a TODO, but as for resolving it here -- not 100% sure how to resolve it. Should we discuss it in more details?

z-a-f replied:
Will post the explanation later -- there was an issue with some functions being compiled, but jit could not handle them.

vkuzo commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -352,11 +381,9 @@ def forward(self, input):
-            _reversed_padding_repeated_twice = _reverse_repeat_padding(self.padding[:1])
-            input = F.pad(input, _reversed_padding_repeated_twice,
-                          mode=self.padding_mode)
+        if isinstance(self.padding, str) or self.padding_mode != 'zeros':
```
in the floating point conv the corresponding check is `if self.padding_mode != 'zeros':`, is there a reason it is different here?

z-a-f replied:
Yes, the FP implementation for the convolution supports different types for convolution. However, in quantization conv (or more accurately in the prepacking cpp code), we don't support taking strings as padding. Hence, we do it here.

z-a-f replied:
One clarification: the FP convolution supports ONLY zero padding (afaik). Basically, that means when calling FP convolution with "reflect" or "circular" padding, it calls the `F.pad` before calling the `F.conv`.

Also, the reason for this additional check is that the FP conv can take string arguments, while quantized conv kernel cannot: internally, the FP conv computes the amount of padding that is required, if the padding is `same`, and only after that calls the conv kernel. In our case, we compute the amount of padding outside the Cpp kernel, and pre-compute it. After that we call F.pad to be able to support "asymmetric" padding (which by design is not supported in conv).

Commit:
Update on "[quant] Enable 'same' and 'valid' padding in static qconv"


Resolves #76304

The string inputs in the quantized convolutions are not implemented, because:
1. Conv packing expects numerical padding as an input
2. The same padding computation is not implemented internally in the quantized conv

To resolve we use a commbination of padding and conv layers:

1. Precompute the padding for the `F.pad` (`_reverse_padding_repeat_twice`)
    1. If `padding == 'same'`: padding[i] = dilation[i] * (kernel[i] - 1)
    2. If `padding == 'valid'`: padding[i] = 0
1. If `padding_mode == 'zeros'` and `padding` is numeric
    1. Prepack the conv with self.padding
    1. Run the qconv in the forward as is
1. If `padding_mode != 'zeros'` or `padding` is a string
    1. In the constructor: Prepack the conv with 0
    1. In the forward:
        1. Run the `F.pad` with `_reversed_padding_repeated_twice`
        1. Run the qconv

The above steps *should not* affect the performance of the current implementaion,
but just add functionality for `'same'` and `'valid'` padding.

Test Plan:

```python
python test/test_quantzation.py TestInputOutputShapes
python test/test_quantzation.py TestStaticQuantizedModule
```

Differential Revision: [D36536582](https://our.internmc.facebook.com/intern/diff/D36536582)

[ghstack-poisoned]

Commit:
Update on "[quant] Enable 'same' and 'valid' padding in static qconv"


Resolves #76304

The string inputs in the quantized convolutions are not implemented, because:
1. Conv packing expects numerical padding as an input
2. The same padding computation is not implemented internally in the quantized conv

To resolve we use a commbination of padding and conv layers:

1. Precompute the padding for the `F.pad` (`_reverse_padding_repeat_twice`)
    1. If `padding == 'same'`: padding[i] = dilation[i] * (kernel[i] - 1)
    2. If `padding == 'valid'`: padding[i] = 0
1. If `padding_mode == 'zeros'` and `padding` is numeric
    1. Prepack the conv with self.padding
    1. Run the qconv in the forward as is
1. If `padding_mode != 'zeros'` or `padding` is a string
    1. In the constructor: Prepack the conv with 0
    1. In the forward:
        1. Run the `F.pad` with `_reversed_padding_repeated_twice`
        1. Run the qconv

The above steps *should not* affect the performance of the current implementaion,
but just add functionality for `'same'` and `'valid'` padding.

Test Plan:

```python
python test/test_quantzation.py TestInputOutputShapes
python test/test_quantzation.py TestStaticQuantizedModule
```

Differential Revision: [D36536582](https://our.internmc.facebook.com/intern/diff/D36536582)

[ghstack-poisoned]

albanD commented on test/allowlist_for_publicAPI.json:
```diff
@@ -2402,6 +2402,7 @@
     "List",
     "Optional",
     "TypeVar",
+    "Union",
```
We should not add new entries to this file.
Adding the proper `__all__` to the submodule should do the trick here.

z-a-f replied:
Got it -- I misunderstood the purpose of this file

Commit:
Update on "[quant] Enable 'same' and 'valid' padding in static qconv"


Resolves #76304

The string inputs in the quantized convolutions are not implemented, because:
1. Conv packing expects numerical padding as an input
2. The same padding computation is not implemented internally in the quantized conv

To resolve we use a commbination of padding and conv layers:

1. Precompute the padding for the `F.pad` (`_reverse_padding_repeat_twice`)
    1. If `padding == 'same'`: padding[i] = dilation[i] * (kernel[i] - 1)
    2. If `padding == 'valid'`: padding[i] = 0
1. If `padding_mode == 'zeros'` and `padding` is numeric
    1. Prepack the conv with self.padding
    1. Run the qconv in the forward as is
1. If `padding_mode != 'zeros'` or `padding` is a string
    1. In the constructor: Prepack the conv with 0
    1. In the forward:
        1. Run the `F.pad` with `_reversed_padding_repeated_twice`
        1. Run the qconv

The above steps *should not* affect the performance of the current implementaion,
but just add functionality for `'same'` and `'valid'` padding.

Test Plan:

```python
python test/test_quantzation.py TestInputOutputShapes
python test/test_quantzation.py TestStaticQuantizedModule
```

Differential Revision: [D36536582](https://our.internmc.facebook.com/intern/diff/D36536582)

[ghstack-poisoned]

Commit:
Update on "[quant] Enable 'same' and 'valid' padding in static qconv"


Resolves #76304

The string inputs in the quantized convolutions are not implemented, because:
1. Conv packing expects numerical padding as an input
2. The same padding computation is not implemented internally in the quantized conv

To resolve we use a commbination of padding and conv layers:

1. Precompute the padding for the `F.pad` (`_reverse_padding_repeat_twice`)
    1. If `padding == 'same'`: padding[i] = dilation[i] * (kernel[i] - 1)
    2. If `padding == 'valid'`: padding[i] = 0
1. If `padding_mode == 'zeros'` and `padding` is numeric
    1. Prepack the conv with self.padding
    1. Run the qconv in the forward as is
1. If `padding_mode != 'zeros'` or `padding` is a string
    1. In the constructor: Prepack the conv with 0
    1. In the forward:
        1. Run the `F.pad` with `_reversed_padding_repeated_twice`
        1. Run the qconv

The above steps *should not* affect the performance of the current implementaion,
but just add functionality for `'same'` and `'valid'` padding.

Test Plan:

```python
python test/test_quantzation.py TestInputOutputShapes
python test/test_quantzation.py TestStaticQuantizedModule
```

Differential Revision: [D36536582](https://our.internmc.facebook.com/intern/diff/D36536582)

[ghstack-poisoned]

Commit:
Update on "[quant] Enable 'same' and 'valid' padding in static qconv"


Resolves #76304

The string inputs in the quantized convolutions are not implemented, because:
1. Conv packing expects numerical padding as an input
2. The same padding computation is not implemented internally in the quantized conv

To resolve we use a commbination of padding and conv layers:

1. Precompute the padding for the `F.pad` (`_reverse_padding_repeat_twice`)
    1. If `padding == 'same'`: padding[i] = dilation[i] * (kernel[i] - 1)
    2. If `padding == 'valid'`: padding[i] = 0
1. If `padding_mode == 'zeros'` and `padding` is numeric
    1. Prepack the conv with self.padding
    1. Run the qconv in the forward as is
1. If `padding_mode != 'zeros'` or `padding` is a string
    1. In the constructor: Prepack the conv with 0
    1. In the forward:
        1. Run the `F.pad` with `_reversed_padding_repeated_twice`
        1. Run the qconv

The above steps *should not* affect the performance of the current implementaion,
but just add functionality for `'same'` and `'valid'` padding.

Test Plan:

```python
python test/test_quantzation.py TestInputOutputShapes
python test/test_quantzation.py TestStaticQuantizedModule
```

Differential Revision: [D36536582](https://our.internmc.facebook.com/intern/diff/D36536582)

[ghstack-poisoned]
